numpy/_core/src/multiarray/dtypemeta.c

```diff
@@ -1230,6 +1233,18 @@ dtypemeta_wrap_legacy_descriptor(
         dtype_class->flags |= NPY_DT_NUMERIC;
     }
 
+    /* If sorting compare not defined, set to arrfunc default */
+    if (dt_slots->sort_compare == NULL) {
+        if (!NPY_DT_is_legacy(dtype_class)) {
+            PyErr_SetString(PyExc_RuntimeError,
+                    "DType has no sort_compare function.");
```

Comment:
I know I said before that we should error for all non-legacy dtypes, but I reconsidered.

If the type sets `NPY_DT_PyArray_ArrFuncs_compare`, couldn't we use that?

Existing user dtypes probably shouldn't error at runtime when this code is added. That said, any non-legacy numpy dtype defined inside numpy should probably trigger this error path.

Reply:
That makes sense; we wouldn't want to surprise existing user dtypes. This is done, thanks!